# Домашнє завдання до теми "Основи роботи з даними у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:
import pandas as pd

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
data_path = 'drive/MyDrive/Data_analitic/Python/Pandas_Python/supermarket_sales.csv'
df = pd.read_csv(data_path)

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [12]:
df.head(4)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4

In [13]:
df.tail(4)

Invoice ID Branch      City Customer type  Gender         Product line  \
996  303-96-2227      B  Mandalay        Normal  Female   Home and lifestyle   
997  727-02-1313      A    Yangon        Member    Male   Food and beverages   
998  347-56-2442      A    Yangon        Normal    Male   Home and lifestyle   
999  849-09-3807      A    Yangon        Member  Female  Fashion accessories   

     Unit price  Quantity  Tax 5%     Total       Date   Time  Payment  \
996       97.38        10  48.690  1022.490   3/2/2019  17:16  Ewallet   
997       31.84         1   1.592    33.432   2/9/2019  13:22     Cash   
998       65.82         1   3.291    69.111  2/22/2019  15:33     Cash   
999       88.34         7  30.919   649.299  2/18/2019  13:28     Cash   

       cogs  gross margin percentage  gross income  Rating  
996  973.80                 4.761905        48.690     4.4  
997   31.84                 4.761905         1.592     7.7  
998   65.82                 4.761905         3.291     4.1  
999  618.38                 4.761905        30.919     6.6

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1F5XlzBOOTMsedpsYqz6NnXjIPTDe6jpx)

2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [16]:
df[615:620]

Invoice ID Branch       City Customer type  Gender  \
615  509-29-3912      A     Yangon        Member  Female   
616  327-40-9673      B   Mandalay        Member    Male   
617  840-19-2096      C  Naypyitaw        Member    Male   
618  828-46-6863      A     Yangon        Member    Male   
619  641-96-3695      C  Naypyitaw        Member  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
615       Sports and travel       38.72         9  17.4240  365.9040   
616       Sports and travel       72.60         6  21.7800  457.3800   
617  Electronic accessories       87.91         5  21.9775  461.5275   
618      Food and beverages       98.53         6  29.5590  620.7390   
619     Fashion accessories       43.46         6  13.0380  273.7980   

          Date   Time      Payment    cogs  gross margin percentage  \
615  3/20/2019  12:24      Ewallet  348.48                 4.761905   
616  1/13/2019  19:51         Cash  435.60                 4.761905   
617  3/14/2019  18:10      Ewallet  439.55                 4.761905   
618  1/23/2019  11:22  Credit card  591.18                 4.761905   
619   2/7/2019  17:55      Ewallet  260.76                 4.761905   

     gross income  Rating  
615       17.4240     4.2  
616       21.7800     6.9  
617       21.9775     4.4  
618       29.5590     4.0  
619       13.0380     8.5

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1Otufe_TXRmjZxWigzBiVBLCAVUfEReXM)

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [22]:
df.dtypes

Invoice ID                  object
Branch                      object
City                        object
Customer type               object
Gender                      object
Product line                object
Unit price                 float64
Quantity                     int64
Tax 5%                     float64
Total                      float64
Date                        object
Time                        object
Payment                     object
cogs                       float64
gross margin percentage    float64
gross income               float64
Rating                     float64
dtype: object

Очікуваний результат (один з можливих форматів):

![](https://drive.google.com/uc?export=view&id=1ASh6-S-_f3X9fTQAPIvnHMFZamTBjNkL)

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [24]:
df['gross income'].describe(percentiles=[0.01, 0.25, 0.5, 0.75, 0.99])

count    1000.000000
mean       15.379369
std        11.708825
min         0.508500
1%          0.913735
25%         5.924875
50%        12.088000
75%        22.445250
99%        45.250750
max        49.650000
Name: gross income, dtype: float64

Данні розміщені в порядку зростання за рівнем прибутковості. 75отму персентилю цієї колонки відповідає 22.445250.  Це означає, що 25% значень в цій колонці є більшими за данне значення. І 75 % значень є меншими. Тобто більша частина значень gross income є меншими за 22.445.  Виходить, що більшість покупок (75%) дають прибуток менший за 22.445. В той час, як менша часта покупок (25%, точніше 24% не беремо 1% потенційних викидів) дають більшу прибутковість.

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1bDa1YFUWDsPm78Lnpy2geQND8M1HEQiT)

5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [27]:
citi_couts = df.City.value_counts()

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1G8c6M875NOW7oU3OtZegTOnyD2y8lhAV)

6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [28]:
citi_couts/citi_couts.sum()

City
Yangon       0.340
Mandalay     0.332
Naypyitaw    0.328
Name: count, dtype: float64

In [36]:
df.City.value_counts(normalize=True)

City
Yangon       0.340
Mandalay     0.332
Naypyitaw    0.328
Name: proportion, dtype: float64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1yZZpWhL2y8DX9rSRioYCRh1VbJgcf3fY)

7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [32]:
df.groupby('City')['gross income'].mean()[['Yangon', 'Mandalay']]

City
Yangon      14.874001
Mandalay    15.232024
Name: gross income, dtype: float64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1JdXUmK46WcoXUJYP7NS2SJlr2XjBbpQ8)

Більший в місті Mandalay = 15.23

### Гіпотези щодо відмінності середнього gross income між містами Yangon та Mandalay

#### Гіпотеза 1: Відмінність у кількості покупок і середній ціні товарів
- Можливо, в Mandalay люди купують більше товарів або дорожчі товари, ніж у Yangon.
- **Як перевірити**:
  - Порахувати кількість транзакцій
  - Порівняти середню ціну товарів або кількість одиниць на чек
  - Порівняти середній чек
- **Як використати**:
  - Стимулювати продаж у Yangon, наприклад, через акції або зміни асортименту

####  Гіпотеза 2: Культурні або економічні особливості регіонів
- Мешканці Yangon і Mandalay можуть відрізнятися за рівнем доходу, стилем життя та споживчою поведінкою, що впливає на обсяг і характер покупок.

**Як перевірити**:
- Залучити зовнішні демографічні дані:
  - Середній дохід населення в регіоні
  - Рівень урбанізації
  - Статистика зайнятості
  - Дані про споживчі звички
- Можна також дослідити розподіл покупок за часом (обідній, вечірній періоди), щоб визначити активність клієнтів.

**Як використати**:
- Адаптувати маркетингову стратегію до специфіки регіону:
  - У місті з великою кількістю офісних працівників пропонувати знижки в обідній та вечірній час для збільшення продажів **готової кулінарії**.
  - У регіонах із більшою кількістю сімей з дітьми — робити акцент на **знижки на дитячі товари**, харчування та побутову продукцію.
  - Формувати асортимент, який краще відповідає очікуванням і стилю життя мешканців кожного регіону.


#### Гіпотеза 3: Демографічна структура покупців впливає на дохід
- В одному місті можуть переважати певні групи покупців (наприклад, жінки, сім'ї з дітьми, неодружені, розлучені), які купують більше або дорожчі товари
- **Як перевірити**:
  - Проаналізувати `Gender`, якщо є — інші дані про сімейний стан або вік (або зібрати їх)
- **Як використати**:
  - Таргетувати найбільш прибуткові сегменти в інших містах за допомогою реклами чи асортименту

#### Гіпотеза 4: Вплив інфраструктури магазинів на купівельну поведінку

 Розташування та комфорт магазинів (кількість магазинів, доступність, умови для покупців) можуть впливати на обсяг продажів у різних містах.

**Як перевірити**:
- Зібрати дані про:
  - Кількість магазинів у кожному місті
  - Їхнє розташування (близькість до житлових районів, офісів, транспортних вузлів)
  - Наявність паркувальних місць
  - Зручність для сімей з дітьми (наявність дитячих кімнат, візочків, кошиків)
  - Політику щодо тварин (чи можна заходити з домашніми улюбленцями)
- Провести опитування клієнтів або аналіз відгуків, щоб оцінити вплив цих факторів

**Як використати**:
- Вдосконалити інфраструктуру магазинів у містах із нижчим `gross income`
- Розвивати додаткові сервіси для підвищення комфорту покупців
- Обирати зручніші локації для відкриття нових магазинів, щоб залучити більше клієнтів



In [52]:
df[df['City'].isin(['Yangon', 'Mandalay'])].groupby('City')['Gender'].value_counts()

City      Gender
Mandalay  Male      170
          Female    162
Yangon    Male      179
          Female    161
Name: count, dtype: int64

In [45]:
df.query("City == 'Yangon'").describe()

Unit price    Quantity      Tax 5%        Total        cogs  \
count  340.000000  340.000000  340.000000   340.000000  340.000000   
mean    54.780853    5.467647   14.874001   312.354031  297.480029   
std     26.132127    2.859876   11.030477   231.640025  220.609547   
min     10.080000    1.000000    0.604500    12.694500   12.090000   
25%     32.250000    3.000000    6.547125   137.489625  130.942500   
50%     53.235000    5.000000   11.468000   240.828000  229.360000   
75%     75.095000    8.000000   21.873375   459.340875  437.467500   
max     99.830000   10.000000   49.490000  1039.290000  989.800000   

       gross margin percentage  gross income      Rating  
count             3.400000e+02    340.000000  340.000000  
mean              4.761905e+00     14.874001    7.027059  
std               1.778975e-14     11.030477    1.731345  
min               4.761905e+00      0.604500    4.000000  
25%               4.761905e+00      6.547125    5.600000  
50%               4.761905e+00     11.468000    7.100000  
75%               4.761905e+00     21.873375    8.500000  
max               4.761905e+00     49.490000   10.000000

In [46]:
df.query("City == 'Mandalay'").describe()

Unit price    Quantity      Tax 5%        Total        cogs  \
count  332.000000  332.000000  332.000000   332.000000  332.000000   
mean    55.659277    5.481928   15.232024   319.872506  304.640482   
std     26.224559    2.896446   11.545122   242.447565  230.902443   
min     10.590000    1.000000    0.887500    18.637500   17.750000   
25%     33.300000    3.000000    5.711250   119.936250  114.225000   
50%     54.850000    5.000000   12.042000   252.882000  240.840000   
75%     77.002500    8.000000   22.170750   465.585750  443.415000   
max     99.960000   10.000000   48.690000  1022.490000  973.800000   

       gross margin percentage  gross income      Rating  
count             3.320000e+02    332.000000  332.000000  
mean              4.761905e+00     15.232024    6.818072  
std               1.601134e-14     11.545122    1.713719  
min               4.761905e+00      0.887500    4.000000  
25%               4.761905e+00      5.711250    5.300000  
50%               4.761905e+00     12.042000    6.700000  
75%               4.761905e+00     22.170750    8.200000  
max               4.761905e+00     48.690000   10.000000

8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [27]:
unique_margins = df['gross margin percentage'].nunique()

if unique_margins == 1:
    print("У всіх інвойсах однакова маржа")
else:
    print("Маржа відрізняється для різних інвойсів")

У всіх інвойсах однакова маржа


Очікуваний результат: відповідь щодо маржі в усіх інвойсах на основі отриманих результатів

9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розпоіділ кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [9]:
import pandas as pd

Очікуваний результат для кожного з міст:

![](https://drive.google.com/uc?export=view&id=1IbG58YK5KVqC-wCnllF1btgEgfWTbi_O)

![](https://drive.google.com/uc?export=view&id=1aKOkLmlMq1sOh4GbwyaQyjfhNJzdU4Yh)

In [39]:
df.groupby('City')['Product line'].value_counts(normalize=True)[['Yangon', 'Mandalay']]

City      Product line          
Yangon    Home and lifestyle        0.191176
          Electronic accessories    0.176471
          Sports and travel         0.173529
          Food and beverages        0.170588
          Fashion accessories       0.150000
          Health and beauty         0.138235
Mandalay  Fashion accessories       0.186747
          Sports and travel         0.186747
          Electronic accessories    0.165663
          Health and beauty         0.159639
          Food and beverages        0.150602
          Home and lifestyle        0.150602
Name: proportion, dtype: float64